File: lib/pacDonations.ts
import { getCache, setCache } from "./cache"
import { getMembersFromDb, getTopPacDonations, type DbMember } from "./db"

export const PAC_DONATIONS_KEY = "pac-donations-v1"
const PAC_DONATIONS_TTL_SECONDS = 6 * 60 * 60
const TOP_N = 1000

export type PacDonationRow = {
  bioguideId: string
  memberName: string
  party: "D" | "R" | "I"
  chamber: "house" | "senate"
  state: string
  pacName: string
  amount: number
}

function normalizeParty(value: string | null | undefined): "D" | "R" | "I" {
  const v = (value ?? "").trim().toUpperCase()
  if (v.startsWith("D")) return "D"
  if (v.startsWith("R")) return "R"
  return "I"
}

export async function getPacDonationLeaderboard(): Promise<PacDonationRow[]> {
  const cached = await getCache<PacDonationRow[]>(PAC_DONATIONS_KEY)
  if (cached && cached.length > 0) return cached

  const [donations, house, senate] = await Promise.all([
    getTopPacDonations(TOP_N),
    getMembersFromDb("house"),
    getMembersFromDb("senate"),
  ])
  if (donations.length === 0) return []

  const byId = new Map<string, DbMember>()
  for (const m of [...house, ...senate]) byId.set(m.bioguide_id, m)

  const rows: PacDonationRow[] = []
  for (const d of donations) {
    const member = byId.get(d.bioguide_id)
    if (!member) continue
    const amount = Number(d.amount)
    if (!Number.isFinite(amount) || amount <= 0) continue
    rows.push({
      bioguideId: d.bioguide_id,
      memberName: member.name,
      party: normalizeParty(member.party),
      chamber: member.chamber === "senate" ? "senate" : "house",
      state: member.state,
      pacName: d.pac_name,
      amount: Math.round(amount),
    })
  }

  rows.sort((a, b) => b.amount - a.amount)
  if (rows.length > 0) await setCache(PAC_DONATIONS_KEY, rows, PAC_DONATIONS_TTL_SECONDS)
  return rows
}

File: backend/app/services/pac_donations.py
"""PAC donations leaderboard (port of lib/pacDonations.ts)."""

import asyncio
import math

from ..core.cache import get_cache, set_cache
from ..core.db import get_members_from_db, get_top_pac_donations

PAC_DONATIONS_KEY = "pac-donations-v1"
PAC_DONATIONS_TTL_SECONDS = 6 * 60 * 60
TOP_N = 1000


def normalize_party(value: str | None) -> str:
    v = (value or "").strip().upper()
    if v.startswith("D"):
        return "D"
    if v.startswith("R"):
        return "R"
    return "I"


async def get_pac_donation_leaderboard() -> list[dict]:
    cached = await get_cache(PAC_DONATIONS_KEY)
    if cached:
        return cached

    donations, house, senate = await asyncio.gather(
        get_top_pac_donations(TOP_N),
        get_members_from_db("house"),
        get_members_from_db("senate"),
    )
    if not donations:
        return []

    by_id = {m["bioguide_id"]: m for m in [*house, *senate]}

    rows: list[dict] = []
    for d in donations:
        member = by_id.get(d["bioguide_id"])
        if not member:
            continue
        try:
            amount = float(d["amount"])
        except (TypeError, ValueError):
            continue
        if not math.isfinite(amount) or amount <= 0:
            continue
        rows.append(
            {
                "bioguideId": d["bioguide_id"],
                "memberName": member["name"],
                "party": normalize_party(member["party"]),
                "chamber": "senate" if member["chamber"] == "senate" else "house",
                "state": member["state"],
                "pacName": d["pac_name"],
                "amount": round(amount),
            }
        )

    rows.sort(key=lambda row: row["amount"], reverse=True)
    if rows:
        await set_cache(PAC_DONATIONS_KEY, rows, PAC_DONATIONS_TTL_SECONDS)
    return rows
